src/jit/lsra.cpp

```diff
@@ -6795,6 +6826,65 @@ void LinearScan::unassignPhysReg(RegRecord* regRec, RefPosition* spillRefPositio
     }
 }
 
+#ifdef _TARGET_ARM_
+//------------------------------------------------------------------------
+// unassignPhysReg: Unassign the given physical register that needs to consider currentInterval
+//
+// Arguments:
+//    regRec                 - the RegRecord to be unasssigned
+//    currentIntervalRegType - the current interval register type
+//
+// Return Value:
+//    None.
+//
+void LinearScan::unassignPhysReg(RegRecord* regRec, RegisterType currentIntervalRegType)
+{
+    RegRecord*   regRecToUnassign         = regRec;
+    Interval*    assignedInterval         = regRecToUnassign->assignedInterval;
+    RefPosition* willBeSpilledRefPosition = assignedInterval->recentRefPosition;
+
+    // We should change the regRecToUnassign if it is part of the double register but not an even register.
+    // or if the current interval has double register but regRecToUnassign is already unassigned.
+    if ((currentIntervalRegType == TYP_DOUBLE && assignedInterval == nullptr) ||
+        (assignedInterval != nullptr && assignedInterval->registerType == TYP_DOUBLE &&
+         !genIsValidDoubleReg(regRec->regNum)))
+    {
+        regRecToUnassign = findAnotherHalfRegRec(regRec);
+        assignedInterval = regRecToUnassign->assignedInterval;
+
+        // another half assignedInterval should not be nullptr.
+        assert(assignedInterval != nullptr);
+        willBeSpilledRefPosition = assignedInterval->recentRefPosition;
+    }
+
+    unassignPhysReg(regRecToUnassign, willBeSpilledRefPosition);
+
+    RegRecord*   anotherRegRec           = nullptr;
+    Interval*    anotherAssignedInterval = nullptr;
+    RefPosition* anotherSpillRefPosition = nullptr;
+
+    // We should check and clear the another float register
+    // if the current interval type is TYP_DOUBLE and the assigned interval type is TYP_FLOAT.
+    if (currentIntervalRegType == TYP_DOUBLE && assignedInterval->registerType == TYP_FLOAT)
+    {
+        anotherRegRec           = findAnotherHalfRegRec(regRecToUnassign);
+        anotherAssignedInterval = anotherRegRec->assignedInterval;
+
+        if (anotherAssignedInterval != nullptr)
+        {
+            anotherSpillRefPosition = anotherAssignedInterval->recentRefPosition;
+            unassignPhysReg(anotherRegRec, anotherSpillRefPosition);
+        }
+    }
```

Comment:
I find the logic here a bit hard to follow, and it is difficult to determine whether all cases are handled. I would suggest structuring it as follows:
```
Interval*  intervalToSpill = regRec->assignedInterval;
RegRecord* secondRegRec = nullptr;

if (newRegType == TYP_DOUBLE)
{
    secondRegRec = findAnotherHalfRegRec(regRec);
}
else if (newRegType == TYP_FLOAT)
{
    if ((intervalToSpill != nullptr) && (intervalToSpill->registerType == TYP_DOUBLE))
    {
        secondRegRec = findAnotherHalfRegRec(regRec);
    }
}
if (intervalToSpill != nullptr)
{
    unassignPhysReg(regRec, (intervalToSpill->recentRefPosition);
}
if ((secondRegRec != nullptr) && (secondRegRec->assignedInterval != nullptr))
{
    unassignPhysReg(secondRegRec, (secondRegRec->assignedInterval->recentRefPosition);
}
```
I think this is both more concise and easier to follow.